src/languages/fsharp.js

```diff
@@ -221,16 +225,134 @@ export default function(hljs) {
     ]
   };
 
-  // 'a or ^a
+  // Most identifiers can contain apostrophes
+  const IDENTIFIER_RE = /[a-zA-Z_](\w|')*/;
+
+  const QUOTED_IDENTIFIER = {
+    scope: 'variable',
+    begin: /``/,
+    end: /``/
+  };
+
+  // 'a or ^a where a can be a ``quoted identifier``
   const GENERIC_TYPE_SYMBOL = {
-    match: regex.concat(/('|\^)/, hljs.UNDERSCORE_IDENT_RE),
+    match: /\B('|\^)(?=``|\w)/,
     scope: 'symbol',
+    contains: [
+      hljs.inherit(QUOTED_IDENTIFIER, {
+        scope: null // Use parent scope
+      }),
+      // Cannot contain another apostrophe
+      hljs.inherit(hljs.UNDERSCORE_TITLE_MODE, {
+        scope: null // Use parent scope
+      })
+    ],
+    relevance: 0
+  };
```

Comment:
I think this has gotten too complex... we're trying to match single identifiers here, yes?  Are there tests for these?

If so, then we probably need a single rule or a `variant` for each perhaps... to match the entire type symbol (quoted or unquoted)... we don't need to rely on `contains` for this as contains can match MULTIPLE times which can lead to false positives in other languages or incorrect parsing:

For example what I think you meant to allow:

- `^[quoted identifier]`
- `^[underscore title]`

But what you actually allow:

```
^[quoted identifier][quoted identifier][underscore_title][quoted identifier]
```

This can be solved with `endsParent` also but I think what we likely need here is just to break out the two cases individually without resorting to children modes.

Reply:
Indeed, you are right. This mode should only match one identifier syntax, prefixed with `'` or '^'.

There are tests in test/markup/fsharp/types.txt and also in other markup tests since it's often used, but I didn't try to break it with invalid edge cases.

If I use `variants` I can't use `begin` in the parent mode though.

Is there a way to reuse the existing `UNDERSCORE_TITLE_MODE` and `QUOTED_IDENTIFIER` and indicate they should be prefixed with another regex? Or did you mean I should actually not reuse these and rewrite a complete regex for both those variants?

Reply:
> Is there a way to reuse the existing UNDERSCORE_TITLE_MODE and QUOTED_IDENTIFIER and indicate they should be prefixed with another regex?

Just reference their attributes directly:

```
QUOTED_IDENTIFIER.match
```

And be consistent with `match/begin`... if there is only a single expression (no end), prefer `match`.

Reply:
> Something like this?:

Exactly.

> but I didn't try to break it with invalid edge cases.

I wasn't suggesting that, only looking for positive examples in a single file, but I didn't see one.